python/lsst/ts/salobj/async_s3_bucket.py

```diff
@@ -0,0 +1,130 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+__all__ = ["AsyncS3Bucket"]
+
+import asyncio
+import io
+
+import boto3
+import botocore
+
+
+class AsyncS3Bucket:
+    """Asynchronous interface to Amazon Web Services s3 buckets.
+
+    Parameters
+    ----------
+    name : `str`
+        Name of bucket.
+    """
+    def __init__(self, name):
+        s3 = boto3.resource("s3")
+        self.name = name
+        self.bucket = s3.Bucket(name)
+
+    async def upload(self, fileobj, key, callback=None):
+        """Upload a file-like object to the bucket.
+
+        Parameters
+        ----------
+        fileobj : file-like object
+            File-like object that can be read as *binary* data.
+        key : `str`
+            Name to use for the file in the bucket.
+        callback : callable (optional)
+            Function to call with updates while writing. The function receives
+            one argument: the number of bytes written. If the transfer is
+            successful then it will always be called at least once,
+            and the sum of the number of bytes for all calls will equal
+            the size of the file.
+
+        Notes
+        -----
+        To create a file-like object from an astropy.io.fits object:
+            fileobj = io.BytesIO()
+            hdulist.writeto(fileobj)
+        """
+        loop = asyncio.get_event_loop()
+        await loop.run_in_executor(None, self._sync_upload, fileobj, key, callback)
+
+    async def download(self, key, callback=None):
+        """Download a file-like object from the bucket.
+
+
+        Parameters
+        ----------
+        key : `str`
+            Name of the file in the bucket.
+        callback : callable (optional)
+            Function to call with updates while writing. The function receives
+            one argument: the number of bytes read so far. If the transfer is
+            successful then it will always be called at least once,
+            and the sum of the number of bytes for all calls will equal
+            the size of the file.
+
+        Returns
+        -------
+        fileobj : `io.BytesIO`
+            The downloaded data as a file-like object.
+        """
+        loop = asyncio.get_event_loop()
+        return await loop.run_in_executor(None, self._sync_download, key, callback)
+
+    async def exists(self, key):
+        """Check if a specified file exists in the bucket.
+
+        Parameters
+        ----------
+        key : `str`
+            Name of the potential file in the bucket.
+        """
+        loop = asyncio.get_event_loop()
+        try:
+            await loop.run_in_executor(None, self._sync_size, key)
+        except botocore.exceptions.ClientError as e:
+            if e.response['Error']['Code'] == '404':
+                return False
+            # Some other error.
```

Comment:
I don't understand the point of this comment? Did you mean "Raise other errors back to the caller."?

Reply:
The only error that means "does not exist" is the one tested above; any other error is unexpected and so re-raised. I tried rewording the comment for clarity:
```
        except botocore.exceptions.ClientError as e:
            if e.response['Error']['Code'] == '404':
                return False
            # Something went wrong; report it.
            raise
```